### DISCUSSION TITLE
Save entry for site

### DISCUSSION THREAD

**Discussion by kennethdenhollander** [craftcms/cms#13675] [Ideas] [answered]
**The problem**
In Craft 4.5.3 we ran into an issue when you want to 'Save entry' you have to fill in all required fields. This is obviously a good feature, however with multiple sites the 'Save entry' action also check's other sites on empty fields. This makes it really confusing for clients using Craft CMS when working with multiple sites where translations in different languages might be added later on. 

I've shared this issue in the Craft CMS discord channel and came to the conclusion that currently, the only work-around is using the automatic saves, disabling or using a console command.

**Feature request**
Default 'save entry' button tries to save all sites which would trigger the empty fields in all sites and display a warning. It would be great to have a default behaviour of this button (or in the dropdown) to 'save only for (current) site' and also have the ability to save it for all sites in case you need it.

Maybe this can even be an option you can configure in the Craft config to set to your personal project preference.

**Comment by thupsi** [comment]
As I understand it, you want to allow entries to have empty `required` fields in other/secondary sites. 

I can think of two options:

- Maybe those fields should not be marked as `required` if they are not critical for an entry to be considered publishable. 
- Entries should be disabled/not published in the secondary sites until all `required` fields are filled.

**Reply by michtio** [reply]
Third option:
- Only the language you adjusted is published, and the others aren't.

**Comment by thupsi** [comment]
Having said that, I think a `propagate field to other sites` field action, would maybe also help. Or a`propagate all fields to empty fields in other sites` entry action?

**Reply by michtio** [reply]
If you don't propagate data across languages or sites. It shouldn't be a blocker if you don't update the others. It's like the errors are now propagated through while content isn't.

**Comment by michtio** [comment]
@thupsi you understand incorrectly. In a country where multiple languages are spoken there are issues with this approach all together, and it's actually a regression rather than an improvement.

Imagine the following:
1. Clients have their main language set to a particular language for the website.  
2. Client has a request --> hey add field X - it should be required for the future
3. I have all my primary language done, now I want to save my primary language and I'm ready to publish, the other language don't matter yet, we'll update these next week. But for legal requirements we at least need to have the English version published!
4. Uh.... we get an error saving now...
5. Wait we can disable the other sites on saving...
6. Uh ... oh no, now the entry in the other language isn't there anymore! We need to enable it again.
7. But my content for that new addition isn't ready yet! So I can't save my entry, unless I disable, or update ...
8. My copywriter can only deliver the content next week, but it really really needs to be online today!
9. Can we make the fields optional? ... Sure!
10. Oh no don't make them optional!!! My content writers add that content in the wrong spot, and they don't fill in the required field!
11... Loop back to the entirety above.


Also propagate field to other sites is not okay at all. In a country where people speak French/Dutch it's an extremely sensitive topic. Serve French content to the Dutch language or vice versa, and you lose trust from your visitors, and they don't return.

Also there are legal requirements that required never to serve different languages through each other. And many more.

**Reply by thupsi** [reply]
@michtio While I'm aware of the points you stated above, I was pointing out the fact that _technically_ Craft is doing the right thing and propagating content is not the same as propagating schema. You are absolutely right though that in practice, from a UX standpoint, and for many users, it can be a regression.

I also agree that having the content ready for all sites/languages should be handled at the organisation/team level and shouldn't be a blocker for you editorial workflow, not from the CMS side at least. So a warning instead of not being able to save the entry would be ideal.

(I apologise if my initial comment came across as dismisive, that wasn't my intention at all.)

**Comment by michtio** [comment]
In general @brandonkelly I think if you don't propagate data through different entries, the error handling should behave the same. If data isn't propagated only the adjusted entry should throw errors, not the others, or at least give it a warning. I love the error messages on top as a warning. But only trigger "site can't be saved" if:

1. Data is propagated through the different sites
2. If you are on a site to be published in that specific site and didn't fill in the required fields.

**Reply by azerozvn** [reply]
Upvote for this since I have a similar issue where I was excited to update to 4.5.3 for [dropdown now support default option correctly](https://github.com/craftcms/cms/pull/13540). It was soon faded when I discovered that I cannot resave entry with these new default option dropdown because Save Entry would trigger save error on another site. And saving on that site would trigger another error loop. 

I was able to resolve it by disable all sites, resave, then re-enable them.

**Comment by brandonkelly** [comment, marked as answer]
It’s not possible to only save an entry for a single site. The only thing we could consider changing is providing a way to revert to pre-4.5 behavior, where we only validate the site that’s getting saved directly. This would reopen a pretty glaring governance issue, where it’s possible to avoid validation errors simply by switching to a different site and saving the entry from there instead. (#11569)

Couple options that might make sense for us to consider:

1. When an entry is already enabled, we can stop validating fields which weren’t actually modified on a draft (so at least the invalid value isn’t a regression, or introducing new invalid content to the site).
2. We could add a config setting for disabling cross-site validation.

We’ll discuss internally…

**Reply by kennethdenhollander** [reply]
Thanks for your reply! I honestly totally get the logic behind the update in 4.5. It makes really much sense if something on an entry doesn't meet the predefined requirements that it shouldn't be saved or published. That's eventually why you define something as 'required'.

The only issue we currently have that it's really confusing for our clients. Sometimes there stuck in a loop of validation errors since you can navigate to another entry with the link in the error. I think just improving the error/validation system with some extra options should also do the job.

**Reply by thupsi** [reply]
@kennethdenhollander I aggree that maybe some options would help. Maybe when there is a validation error on fields from other sites, there could be an option to `copy value from current site`, or `ignore this validation error`.

**Reply by brandonkelly** [reply]
Craft 4.5.6 is out with that first idea implemented: we no longer validate cross-site field values that weren’t actually modified by the save (regardless of whether the entry is enabled).

**Reply by MoritzLost** [reply]
@brandonkelly Can you clarify when a cross-site validation might present errors, or how to prevent this as a developer and/or editor?

I've got clients that are having issues with this. They edit contents in the primary language and get errors because of empty fields in another language. But I can't reliably reproduce this, when I create a new entry I don't get those cross-site errors, even when fields are empty in other languages. In what scenario would the cross-site validation prevent an entry from being saved?

The current behaviour is definitely not ideal for my clients. In a lot of cases, they have a primary language and secondary language, and they want to enter content in the primary language freely, and then worry about the secondary language later. But the cross-site validation makes this difficult.

**Reply by brandonkelly** [reply]
@MoritzLost When you were testing it yourself, were you actually saving the entries as enabled? Because required field validation will only kick in when the entry is supposed to be live.

> The current behaviour is definitely not ideal for my clients. In a lot of cases, they have a primary language and secondary language, and they want to enter content in the primary language freely, and then worry about the secondary language later. But the cross-site validation makes this difficult.

It sounds like either you should change the Default Status settings to Disabled, so only the site that was intentionally edited will get strictly validated and published, or loosen up your validation requirements if you don’t actually care about these fields being filled in.

**Reply by MoritzLost** [reply]
@brandonkelly 

> When you were testing it yourself, were you actually saving the entries as enabled? Because required field validation will only kick in when the entry is supposed to be live.

Yes, I've tested with a live entry. Here's an example of this behaviour:

https://github.com/craftcms/cms/assets/10146880/45d3291c-9132-4087-840e-606338d05200

The matrix block contains a Super Table field with two required  fields inside it (Asset and Typed Link). When I create a new Super Table row and leave both fields empty, I correctly get a validation error. Then I fill out those fields and save the entry again. At this point I _don't_ get a cross-site error, even though the value is still empty in the English site. When I save the English site entry itself, I get an error. Not sure if that's an issue in the core or in the plugin fields.

This is what happens when I try to reproduce the issue with the unexpected cross-site validation error. However, my clients _are_ getting validation errors in a similar scenario. I told them that it's only happening with existing contents that were created before the update to the latest Craft version. But they're still getting this error even with new entries.

> It sounds like either you should change the Default Status settings to Disabled, so only the site that was intentionally edited will get strictly validated and published, or loosen up your validation requirements if you don’t actually care about these fields being filled in.

Yeah, I'm thinking about either that or setting those sections to manual propagation, so the clients can decide themselves when to propagate their entries to English. But both of those are a bit more work for the editors, and I'm still trying to find whether this is only a 'legacy' issue for content create before the validation changes, or can actively happen for new entries in some way.

**Reply by brandonkelly** [reply]
> But they're still getting this error even with new entries.

Are you able to reproduce that? Same field setup?

**Reply by MoritzLost** [reply]
@brandonkelly Yes, they've saved a version of an entry where this issue occurs as a draft, I've pulled it down locally to investigate. When I try to apply the draft, the same field throws validation errors in the other site:


https://github.com/craftcms/cms/assets/10146880/aadb0a33-527a-4301-8042-be011b37c77e

**Reply by brandonkelly** [reply]
@MoritzLost In that video, the Matrix field is marked as changed on both sites. Do you know whether the block in question was added/edited within this draft before you started recording?

**Reply by brandonkelly** [reply]
I’m testing with the same setup locally, as far as I can tell from your video:

- Two sites
- Matrix field (with Propagation Method set to “Save blocks to all sites the owner element is saved in”)
- Nested Assets field (required; “Manage relations on a per-site basis” is _disabled_)
- Nested Typed Link field (required; Translation Method set to “Translate for each site”)

I initially marked those nested Assets and Typed Link fields as _not required_, and created a couple Matrix blocks and nested Super Table rows with values filled-in for Site A but not Site B. Then I marked those nested fields as required.

So now, saving Site B results in required validation errors on the Typed Link fields, as expected.

However when I exclusively edit the entry from Site A, nothing I do will trip validation errors on Site B for those fields.

I’ve tried:

- editing an existing Typed Link value in Site A, whose corresponding Site B value is still empty;
- creating a new Super Table row in Site A, but not setting its value in Site B (this works because the Site A Typed Link value ends up getting copied over to Site B automatically, since it’s a freshly-created Super Table row);
- adding a new Matrix block + Super Table row in Site A, but not setting its values in Site B.

No matter what I try, everything is saving without cross-site validation errors in Site B, because I’m not editing Site B’s values directly.

**Reply by MoritzLost** [reply]
@brandonkelly 

> In that video, the Matrix field is marked as changed on both sites. Do you know whether the block in question was added/edited within this draft before you started recording?

No, the block already existed in the canonical entry. I compared the draft with the canonical entry in both sites. The matrix block exists in both sites in both the draft and the canonical entry. The values are also identical between the draft and the canonical entry. In German, there are four super table blocks with both fields filled out. In English, there are the same four super table entries, but the typed link fields are _empty_. But the draft and canonical version are identical.

---

Your test setup is correct, that matches our settings exactly. Same as you, I can't reproduce the problem on a fresh entry, only on the existing entries that were created before updating Craft. Hm, maybe it's really only an issue with old entries. The client told me it also happens with new entries, but I've checked the other occurrences again, and those are entries that were created before the last Craft update (that introduced cross-site validation). So maybe it really is only an issue for old contents that could be published before the cross-site validation was introduced, and therefore have validation errors in other sites?

What I still don't understand is when the cross-site validation would actually take place. In my testing, when creating a new entry, the cross-site validation was never triggered as long as I only edit the German site, even if required fields are left empty in English. In what scenario would a cross-site validation prevent the German entry from saving? Can this happen if the editor only ever touches the German site (so the English contents only include the initial propagated field values)?

**Reply by brandonkelly** [reply]
As of 4.5.6+, cross-site validation will only occur for fields which were directly edited on the other site(s). So if your editor goes in to edit an entry in German, and never switches to English, it shouldn’t ever trigger a cross-site validation error.

In your video, the draft did already show a record of the field being changed across both sites. I’m not sure how it got into that state, but that’s why you get the cross-site validation error.

**Reply by MoritzLost** [reply]
@brandonkelly Thanks for the clarification! Maybe my client accidentally changed something in the English and forgot about it, or it was some sort of edge-case coming from the update. I'll let them know again how to avoid that issue, and hopefully it will be smooth sailing from here on out ⛵

**Comment by angelokeirsebilck** [comment]
Did you already discuss this internally? Any update on the decision?

**Reply by brandonkelly** [reply]
Yep, we decided to only cross-site validate fields that were modified (c223e9012a74ba275d72f3af5749dbc315b4ef7c). Update coming soon…
